File: objc-cocoapods/objc-cocoapods/ViewController.m
//
//  ViewController.m
//  objc-cocoapods
//
//  Created by k@binc.jp on 10/20/16.
//  Copyright © 2016 PAY.JP. All rights reserved.
//

#import "ViewController.h"
#import "AppDelegate.h"

@import PassKit;
@import PAYJP;

@interface ViewController () <PKPaymentAuthorizationViewControllerDelegate>

@property (nonatomic, weak) IBOutlet UITextField *amountField;
@property (nonatomic, weak) IBOutlet UITextView *responseText;
@property (nonatomic, weak) IBOutlet UIView *buttonArea;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    UIButton *button = [PKPaymentButton buttonWithType:PKPaymentButtonTypeBuy style:PKPaymentButtonStyleBlack];
    [button addTarget:self action:@selector(pay) forControlEvents:UIControlEventTouchUpInside];
    [self.buttonArea insertSubview:button atIndex:0];
}

- (void)pay {
    PKPaymentRequest *request = [[PKPaymentRequest alloc] init];
    request.currencyCode = @"JPY";
    request.countryCode = @"JP";
    request.merchantIdentifier = ApplePayMerchantID;
    
    NSDecimalNumber *amount = [NSDecimalNumber decimalNumberWithString:self.amountField.text];
    PKPaymentSummaryItem *item = [PKPaymentSummaryItem summaryItemWithLabel:@"PAY.JP TEST ITEM" amount:amount];
    request.paymentSummaryItems = @[item];
    request.supportedNetworks = [PKPaymentRequest availableNetworks];
    request.merchantCapabilities = PKMerchantCapability3DS;
    request.requiredBillingAddressFields = PKAddressFieldPostalAddress;
    
    PKPaymentAuthorizationViewController *viewController = [[PKPaymentAuthorizationViewController alloc] initWithPaymentRequest:request];
    viewController.delegate = self;
    [self presentViewController:viewController animated:YES completion:NULL];
}

#pragma mark - PKPaymentAuthorizationViewControllerDelegate

- (void)paymentAuthorizationViewController:(PKPaymentAuthorizationViewController *)controller
                       didAuthorizePayment:(PKPayment *)payment
                                completion:(void (^)(PKPaymentAuthorizationStatus status))completion {
    
    PAYAPIClient *apiClient = PAYAPIClient.sharedClient;
    [apiClient createTokenWith:payment.token completionHandler:^(PAYToken *token, NSError *error) {
        if (!token) {
            completion(PKPaymentAuthorizationStatusFailure);
            return;
        }
        
        NSLog(@"%@", token);
        completion(PKPaymentAuthorizationStatusSuccess);
    }];
}

- (void)paymentAuthorizationViewControllerDidFinish:(PKPaymentAuthorizationViewController *)controller {
    NSLog(@"call delegate paymentAuthorizationViewControllerDidFinish:");
    [controller dismissViewControllerAnimated:YES completion:NULL];
}

@end

File: swift-carthage/swift3-carthage/ViewController.swift
import UIKit
import PassKit

import PAYJP

class ViewController: UIViewController, PKPaymentAuthorizationViewControllerDelegate {
    @IBOutlet weak var amountField: UITextField!
    @IBOutlet weak var responseText: UITextView!
    @IBOutlet weak var buttonArea: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        var button: UIButton
        if #available(iOS 9.0, *) {
            button = PKPaymentButton(paymentButtonType: .buy, paymentButtonStyle: .black)
        } else {
            button = UIButton(frame: buttonArea.frame)
            let image = UIImage(named: "ApplePayBTN_32pt__black_textLogo_")
            button.setImage(image, for: .normal)
        }
        
        button.addTarget(self, action: #selector(pay), for: UIControl.Event.touchUpInside)
        buttonArea.insertSubview(button, at: 0)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if checkApplePayAvaliable() {
            checkPaymentNetworksAvaliable(usingNetworks: supportedPaymentNetworks)
        }
    }

    func paymentAuthorizationViewController(_ controller: PKPaymentAuthorizationViewController, didAuthorizePayment payment: PKPayment, completion: @escaping (PKPaymentAuthorizationStatus) -> Void) {
        guard let amount = currentAmount?.stringValue else { return }
        
        let apiClient = APIClient.shared
        apiClient.createToken(with: payment.token) { (result) in
            switch result {
            case .success(let token):
                let card = token.card
                print("pay amount: \(amount), card: ***\(card.last4Number) token: \(token.identifer)")
                completion(.success)
                // handle token object and send back to your server.
                // https://pay.jp/docs/charge
            case .failure(let error):
                print(error)
                completion(.failure)
            }
        }
    }

    func paymentAuthorizationViewControllerDidFinish(_ controller: PKPaymentAuthorizationViewController) {
        controller.dismiss(animated: true, completion: nil)
    }
    
    @objc func pay() {
        guard let amount = currentAmount else {
            return
        }
        
        // make an ApplePay request
        let request = PKPaymentRequest()
        request.currencyCode = "JPY"
        request.countryCode = "JP"
        request.merchantIdentifier = appleMerchantID
        
        let item = PKPaymentSummaryItem(label: "PAY.JP TEST ITEM", amount: amount)
        request.paymentSummaryItems = [item]
        request.supportedNetworks = supportedPaymentNetworks
        request.merchantCapabilities = PKMerchantCapability.capability3DS
        request.requiredBillingAddressFields = PKAddressField.postalAddress
        
        if let vc = PKPaymentAuthorizationViewController(paymentRequest: request) {
            vc.delegate = self
            present(vc, animated: true, completion: nil)
        }
    }
    
    private var supportedPaymentNetworks: [PKPaymentNetwork] {
        get {
            if #available(iOS 10.0, *) {
                return PKPaymentRequest.availableNetworks()
            } else {
                return [.visa, .masterCard, .amex]
            }
        }
    }
    
    private var currentAmount: NSDecimalNumber? {
        get { return NSDecimalNumber(string: amountField.text) }
    }
    
    private func checkApplePayAvaliable() -> Bool {
        if !PKPaymentAuthorizationViewController.canMakePayments() {
            let alertController = UIAlertController(
                title: "エラー",
                message: "このデバイスはApple Payに対応していません",
                preferredStyle: .alert
            )
            alertController.addAction(UIAlertAction(title: "OK", style: UIAlertAction.Style.cancel, handler: nil))
            self.present(alertController, animated: true, completion: nil)
            
            buttonArea.isHidden = true
            
            return false
        }
        return true
    }
    
    private func checkPaymentNetworksAvaliable(usingNetworks networks: [PKPaymentNetwork]) {
        if !PKPaymentAuthorizationViewController.canMakePayments(usingNetworks: networks) {
            let alertController = UIAlertController(
                title: "決済方法が登録されていません",
                message: "今すぐ決済方法を登録しますか？",
                preferredStyle: .alert
            )
            alertController.addAction(UIAlertAction(title: "YES", style: UIAlertAction.Style.default, handler: { action in
                if #available(iOS 8.3, *) {
                    PKPassLibrary().openPaymentSetup()
                }
            }))
            alertController.addAction(UIAlertAction(title: "NO", style: UIAlertAction.Style.cancel, handler: nil))
            self.present(alertController, animated: true, completion: nil)
        }
    }
}



